fury/ui/elements.py

```diff
@@ -3052,3 +3055,394 @@ def directory_click_callback(self, i_ren, _obj, listboxitem):
+        mode_panel_size = (len(mode_data) * 35 + 2 * padding, 40)
+        self.mode_panel = Panel2D(size=mode_panel_size, color=(0.5, 0.5, 0.5))
+        btn_pos = np.array([0, 0])
+
+        for mode, fname in mode_data.items():
+            icon_files = []
+            icon_files.append((mode, read_viz_icons(style="new_icons", fname=fname[0])))
+            icon_files.append((mode+"-pressed", read_viz_icons(style="new_icons", fname=fname[1])))
+            btn = Button2D(icon_fnames=icon_files)
+
+            def mode_selector(i_ren, _obj, btn):
+                self.current_mode = btn.icon_names[0]
+                i_ren.force_render()
+
+            btn.on_left_mouse_button_pressed = mode_selector
+
+            self.mode_panel.add_element(btn, btn_pos+padding)
+            btn_pos[0] += btn.size[0]+padding
+
+        self.canvas.add_element(self.mode_panel, (0, 0))
+
+        self.mode_text = TextBlock2D(text="Select appropriate drawing mode using below icon")
+        self.canvas.add_element(self.mode_text, (0.0, 0.95))
+
+    def _get_actors(self):
+        """Get the actors composing this UI component."""
+        return self.canvas.actors
+
+    def _add_to_scene(self, scene):
+        """Add all subcomponents or VTK props that compose this UI component.
+
+        Parameters
+        ----------
+        scene : scene
+
+        """
+        self.current_scene = scene
+        self.canvas.add_to_scene(scene)
+
+    def _get_size(self):
+        return self.canvas.size
+
+    def _set_position(self, coords):
+        """Set the lower-left corner position of this UI component.
+
+        Parameters
+        ----------
+        coords: (float, float)
+            Absolute pixel coordinates (x, y).
+        """
+        self.canvas.position = coords
+
+    def resize(self, size):
+        """Resize the UI.
+        """
+        pass
+
+    @property
+    def current_mode(self):
+        return self._current_mode
+
+    @current_mode.setter
+    def current_mode(self, mode):
+        self.update_button_icons(mode)
+        self._current_mode = mode
+        if mode is not None:
+            self.mode_text.message = f"Mode: {mode}"
+
+    def cal_min_boundary_distance(self, position):
+        """Calculate the minimum distance between the current position and canvas boundary.
+
+        Parameters
+        ----------
+        position: (float,float)
+            current position of the shape.
+
+        Returns
+        -------
+        float
+            Minimum distance from the boundary.
+        """
+        distance_list = []
+        # calculate distance from element to left and lower boundary
+        distance_list.extend(position - self.canvas.position)
+        # calculate distance from element to upper and right boundary
+        distance_list.extend(self.canvas.position + self.canvas.size - position)
+
+        return min(distance_list)
+
+    def draw_shape(self, shape_type, current_position, in_process=False):
```

Comment:
Maybe I should rename this as `is_drawing` because we are basically checking whether the shape is being drawn.
The job of this parameter is to indicate whether the shape is to be drawn (`in_process` is false viz. shape drawing is not in process) or the shape is already drawn we just have to resize it (`in_progress` is true viz. the shape drawing is in process).

Also, you are getting the _IndexError_ because the shape isn't created yet. As we are saying the shape creation is in process (`in_process` = True) there should exist an actor in the `actor_list` which is not in this case.